subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/dependency_verification.adoc

```diff
@@ -0,0 +1,849 @@
+
+By default, Gradle will not only verify artifacts (jars, ...) but also the metadata associated with those artifacts (typically POM files).
+Verifying this ensures the maximum level of security: metadata files typically tell what transitive dependencies will be included, so a compromised metadata file may cause the introduction of undesired dependencies in the graph.
+However, because all artifacts are verified, such artifacts would in general easily be discovered by you, because they would cause a checksum verification failure (checksums would be _missing_ from verification metadata).
+Because metadata verification can significantly increase the size of your configuration file, you may therefore want to disable verification of metadata.
+If you understand the risks of doing so, set the `<verify-metadata>` flag to `false` in the configuration file:
+
+[source,xml]
+----
+<?xml version="1.0" encoding="UTF-8"?>
+<verification-metadata>
+   <configuration>
+      <verify-metadata>true</verify-metadata>
+      <verify-signatures>false</verify-signatures>
+    </configuration>
+    <!-- the rest of this file doesn't need to declare anything about metadata files -->
+</verification-metadata>
+----
+
+[[sec:checksum-verification]]
+== Verifying dependency checksums
+
+Checksum verification allows you to ensure the integrity of an artifact.
+This is the simplest thing that Gradle can do for you to make sure that the artifacts you use are un-tampered.
+
+Gradle supports MD5, SHA1, SHA-256 and SHA-512 checksums.
+However, only SHA-256 and SHA-512 checksums are considered secure nowadays.
+
+=== Adding the checksum for an artifact
+
+External components are identified by GAV coordinates, then each of the artifacts by their file names.
+To declare the checksums of an artifact, you need to add the corresponding section in the verification metadata file.
+For example, to declare the checksum for https://pdfbox.apache.org[Apache PDFBox].
+The GAV coordinates are:
+
+- group `org.apache.pdfbox`
+- name `pdfbox`
+- version `2.0.17`
+
+Using this dependency will trigger the download of 2 different files:
+
+- `pdfbox-2.0.17.jar` which is the main artifact
+- `pdfbox-2.0.17.pom` which is the metadata file associated with this artifact
+
+As as consequence, you need to declare the checksums for both of them (unless you <<sec:disabling-metadata-verification,disabled metadata verification>>):
+
+[source,xml]
+----
+<?xml version="1.0" encoding="UTF-8"?>
+<verification-metadata>
+   <configuration>
+      <verify-metadata>true</verify-metadata>
+      <verify-signatures>false</verify-signatures>
+   </configuration>
+   <components>
+      <component group="org.apache.pdfbox" name="pdfbox" version="2.0.17">
+         <artifact name="pdfbox-2.0.17.jar">
+            <sha512 value="7e11e54a21c395d461e59552e88b0de0ebaf1bf9d9bcacadf17b240d9bbc29bf6beb8e36896c186fe405d287f5d517b02c89381aa0fcc5e0aa5814e44f0ab331" origin="PDFBox Official site"/>
```

Comment:
It would probably make sense to add a link to `PDFBox Official site`, so end users are encouraged to use verifiable links. An alternative option is to add `origin-url`.

Reasoning:
Suppose someone suggests a pull request that adds `sha512.... origin="PDFBox Official site"`

How should the maintainer verify if the checksum is indeed published on the official site?
If the PR uses `origin="PDFBox Official site: https://pdfbox..../releases/2.0.7..."`, then it is much easier for maintainer to verify the checksum.
Having `<sha512 ... origin="..." origin-url="https://..."` would probably make the file machine-readable (e.g. people might enforce that every checksum must be accompanied with origin-url)

Reply:
Yeah I'm probably biased by the fact I want to avoid storing too much information in memory. You can put whatever you want in "origin". I wouldn't necessarily encourage another field for this.